Bring docs in sync with inertia-vue

MichaelDeBoey opened the pull request:
Just like inertiajs/inertia-vue@9a38473

According to [this](https://michaelnthiessen.com/vue-props-vs-data/) article (and many others), `data` is local state:
> Data is the private memory of each component where you can store any variables you need. Props are how you pass this data from a parent component down to a child component.

With this in mind, I converted the examples from the `inertia-vue` README to using the `useState` hook.

@reinink If I'm correct the second example uses the `user` prop?  
If you want, I can make that a bit clearer here and use `props.user[key]` instead of just `user[key]` in the second example.

MichaelDeBoey commented:
@reinink If you'd like me to reflect some wording changes you've made in inertiajs/inertia-vue@ 1c160c8, I'm happy to change them in this PR too. 🙂

reinink commented:
By all means! I just tweaked that wording a little further.

I will defer to @sebastiandedeyne on this PR though, since I'm honestly not exactly sure how the final implementation in React should look.

Thanks so much for all your help Michael!

MichaelDeBoey commented:
@reinink No problem at all! 🙂

Will tweak the wording a bit like you did in the `inertia-vue` repo.

You like to have the manual visit comments be deleted too?
React doesn't have a plugin system like Vue has, so that's not something I can copy over and implement tho

reinink commented:
Nah, I'd say keep that section as is. In React you'll just have to import Inertia...like we have to this point. 👍

MichaelDeBoey commented:
Wording should be fine now 🙂

reinink commented:
I'm not seeing the changes?

MichaelDeBoey commented:
Oh I squashed my commits and rebased I'm sorry.

What about the second example I mentioned above btw?
> @reinink If I'm correct the second example uses the `user` prop?  
> If you want, I can make that a bit clearer here and use `props.user[key]` instead of just `user[key]` in the second example.

reinink commented:
> What about the second example I mentioned above btw?

I'm not really sure what you mean TBH, show me an example in a comment?

MichaelDeBoey commented:
I mean this one from the `inertia-vue` docs

> If your page contains multiple components using the remember functionality, you'll need to provide a unique key for each component.
> 
> ```js
> data() {
>   return {
>     form: Inertia.remember({
>       first_name: this.user.first_name,
>       last_name: this.user.last_name,
>       // ...
>     }, `Users/Edit:${this.user.id}`),
>   }
> },
> ```

I assume the `this.user` is gotten from the component's  props?

reinink commented:
> I assume the this.user is gotten from the component's props?

Yes, that's correct. I figured, at least in the Vue docs, that this is kind of assumed.

reinink commented:
It's also worth mentioning that the remember functionality now uses two methods:

```js
Inertia.remember(data, key)
Inertia.restore(key)
```

I suspect @sebastiandedeyne will have to update the library accordingly.

MichaelDeBoey commented:
> I figured, at least in the Vue docs, that this is kind of assumed.

What I've read about Vue, it's logical indeed.  
But I think it would be a bit clearer to change the React docs to use `props` explicitly?

```js
const [formState, setFormState] = useState(Inertia.remember({
  first_name: props.user.first_name,
  last_name: props.user.last_name,
  // ...
}, `Users/Edit:${props.user.id}`))
```

MichaelDeBoey commented:
>  I suspect @sebastiandedeyne will have to update the library accordingly.

The Vue and React libraries don't implement anything for the `Inertia.remember` functionality, so I'm not sure if it's needed for the `Inertia. restore` functionality?  
Only updating the docs I think?

reinink commented:
It's coming. Sebastian has some work to do in this regard, and I added it to the Vue adapter today.

https://github.com/inertiajs/inertia-vue/blob/master/src/remember.js#L25

MichaelDeBoey commented:
Happy to help too btw 🙂

But back to this PR: add the `props` key or not?

sebastiandedeyne commented on readme.md:
```diff
@@ -188,22 +165,22 @@ The easiest way to access page props is with our `usePageProps` hook.
 import { InertiaLink, usePageProps } from 'inertia-react'
 
-function Layout({ children }) {
+export default ({ children }) => {
```
Please keep named functions for components. `export default` is generally a bad idea (see https://twitter.com/dan_abramov/status/1093901838179618816)

sebastiandedeyne commented on readme.md:
```diff
@@ -215,23 +192,47 @@ If you need to access the entire Inertia `page` object, you can directly access
 import { InertiaLink, PageContext } from 'inertia-react'
 
-function Layout({ children }) {
+export default ({ children }) => {
```
Same as above

sebastiandedeyne commented on readme.md:
```diff
@@ -215,23 +192,47 @@ If you need to access the entire Inertia `page` object, you can directly access
+To mitigate this issue, you can use `Inertia.remember()` to tell Inertia.js which local component state to cache.
+
+~~~js
+const [formState, setFormState] = useState(Inertia.remember({
```
This doesn't work, so this paragraph can be left out until the hook is released (soon)

MichaelDeBoey replied:
You mean only the example or the whole "Remembering local component state" section?

MichaelDeBoey commented:
@sebastiandedeyne Made the changes you asked for

MichaelDeBoey commented:
@sebastiandedeyne Updated with mentioning the new `useRememberedState` hook

reinink commented:
I'll let @sebastiandedeyne merge this in if he's satisfied. Thanks @MichaelDeBoey!

sebastiandedeyne commented on readme.md:
```diff
@@ -185,53 +165,82 @@ Sometimes it's necessary to access the page data (props) from a non-page compone
+import React from 'react'
 
-function Layout({ children }) {
+const Layout = ({ children }) => {
```
Could you please use the `function` keyword? It's how the rest of the codebase is set up :)

sebastiandedeyne commented on readme.md:
```diff
@@ -185,53 +165,82 @@ Sometimes it's necessary to access the page data (props) from a non-page compone
 
-~~~js
-import { useContext } from 'react'
+~~~jsx harmony
```
Does this do anything different than `js`? I'd rather keep things simple

MichaelDeBoey replied:
It's for correct code highlighting

MichaelDeBoey commented:
@sebastiandedeyne Done as requested 🙂

sebastiandedeyne commented:
Thanks!

MichaelDeBoey commented:
With pleasure! 🙂
